src/Sarif/Visitors/RunMergingVisitor.cs

```diff
@@ -78,7 +78,23 @@ public override Result VisitResult(Result node)
             }
 
             // Cache RuleId and set Result.RuleIndex to the (new) index
-            string ruleId = node.ResolvedRuleId(CurrentRun);
+            string resultRuleId = node.ResolvedRuleId(CurrentRun);
+            string ruleId = resultRuleId;
+            ReportingDescriptor rule = null;
+
+            // Result's rule Id is a sub rule Id
+            if (!string.IsNullOrEmpty(ruleId) &&
+                ruleId.IndexOf(SarifConstants.HierarchicalComponentSeparator) > 0)
+            {
+                rule = node.GetRule(CurrentRun);
+                string possilbeParentRuleId = rule?.Id;
+                if (possilbeParentRuleId != null &&
+                    ruleId.StartsWith(possilbeParentRuleId + SarifConstants.HierarchicalComponentSeparator))
+                {
+                    ruleId = possilbeParentRuleId;
+                }
+            }
```

Comment:
can't we replace this for a smaller logic?

ReportingDescriptor rule = result.GetRule(CurrentRun);
use rule everywhere when possible.

Reply:
agree. The additional checks is to make sure the case is a rule with sub rule. I can trust result.GetRule(...) and remove these checks.

Reply:
updated with simple implementation